### PR TITLE
Fix D support on Windows

### PR DESCRIPTION
**Pull request by GoaLitiuM**
Support for D projects on Windows did not seem to work out of box, so I made the following changes to fix the problem:
- ~~D compilers does not require static linker~~
- Fixed DMD always compiling to 32-bit targets on 64-bit hosts (this was the default behaviour of DMD on Windows)
- If MSVC environment variables are present, compilers should use the target setup by the MSVC toolchain. Similarly how C/C++ VS projects are built with meson.
- Fixed sanity check not using the correct linker flags to build for native architecture

This fixes the issue #3979 I submitted earlier.

### CODE REVIEW

File: mesonbuild/compilers/d.py [hunk 1]
```diff
@@ -42,28 +42,32 @@
...
+    def __init__(self, exelist, version, is_cross, is_64, **kwargs):
         self.language = 'd'
         super().__init__(exelist, version, **kwargs)
         self.id = 'unknown'
         self.is_cross = is_cross
+        self.is_64 = is_64
 
     def sanity_check(self, work_dir, environment):
+        return self.sanity_check_impl(work_dir, environment, [])
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[W293] blank line contains whitespace

File: mesonbuild/compilers/d.py [hunk 2]
```diff
@@ -290,10 +305,13 @@ def build_rpath_args(self, build_dir, from_dir, rpath_paths, build_rpath, instal
 
 
 class LLVMDCompiler(DCompiler):
-    def __init__(self, exelist, version, is_cross, **kwargs):
-        DCompiler.__init__(self, exelist, version, is_cross, **kwargs)
+    def __init__(self, exelist, version, is_cross, is_64, **kwargs):
+        DCompiler.__init__(self, exelist, version, is_cross, is_64, **kwargs)
         self.id = 'llvm'
         self.base_options = ['b_coverage', 'b_colorout']
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[W293] blank line contains whitespace

File: mesonbuild/compilers/d.py [hunk 3]
```diff
@@ -384,10 +408,24 @@ def get_linker_search_args(self, dirname):
         return ['-L-L' + dirname]
 
     def get_buildtype_args(self, buildtype):
+        if buildtype != 'plain':
+            return self.get_target_arch_args() + d_dmd_buildtype_args[buildtype]
         return d_dmd_buildtype_args[buildtype]
 
     def get_std_shared_lib_link_args(self):
         return ['-shared', '-defaultlib=libphobos2.so']
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[W293] blank line contains whitespace

File: mesonbuild/compilers/d.py [hunk 4]
```diff
@@ -196,6 +202,15 @@ def compiles(self, code, env, extra_args=None, dependencies=None, mode='compile'
 
     def has_multi_arguments(self, args, env):
         return self.compiles('int i;\n', env, extra_args=args)
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[W293] blank line contains whitespace

File: mesonbuild/environment.py [hunk 5]
```diff
@@ -834,12 +834,23 @@ def detect_d_compiler(self, want_cross):
             raise EnvironmentException('Could not execute D compiler "%s"' % ' '.join(exelist))
         version = search_version(out)
         full_version = out.split('\n', 1)[0]
+
```
Comment (ignatenkobrain) [inline mesonbuild/environment.py]: __[Flake8]__

[W293] blank line contains whitespace

File: mesonbuild/environment.py [hunk 6]
```diff
@@ -834,12 +834,23 @@ def detect_d_compiler(self, want_cross):
...
+        is_win64 = False
+        is_msvc = False
+        if mesonlib.is_windows():
+            is_msvc = 'VCINSTALLDIR' in os.environ
+            # detect which MSVC build environment is currently active
+            if is_msvc and detect_cpu_family({'c': self.detect_c_compiler(want_cross)}) == 'x86_64':
+                is_win64 = True
+            elif not is_msvc and detect_cpu_family({}) == 'x86_64':
+                is_win64 = True
```
Comment (nirbheek) [inline mesonbuild/environment.py]: We already have code to do this detection, see: `detect_windows_arch()`. Does that not work?

File: mesonbuild/environment.py [hunk 7]
```diff
@@ -834,12 +834,23 @@ def detect_d_compiler(self, want_cross):
...
+
         if 'LLVM D compiler' in out:
-            return compilers.LLVMDCompiler(exelist, version, is_cross, full_version=full_version)
+            return compilers.LLVMDCompiler(exelist, version, is_cross, is_win64, full_version=full_version)
         elif 'gdc' in out:
-            return compilers.GnuDCompiler(exelist, version, is_cross, full_version=full_version)
+            return compilers.GnuDCompiler(exelist, version, is_cross, is_win64, full_version=full_version)
         elif 'The D Language Foundation' in out or 'Digital Mars' in out:
-            return compilers.DmdDCompiler(exelist, version, is_cross, full_version=full_version)
+            return compilers.DmdDCompiler(exelist, version, is_cross, is_win64, is_msvc, full_version=full_version)
```
Comment (nirbheek) [inline mesonbuild/environment.py]: You can't use the LLVM D compiler with MSVC?

**Comment by nirbheek** [comment]
Test failure seems to be relevant. This PR breaks building of static libraries with D.

Also, once everything is fixed and done, this will need a release notes snippet, see: https://mesonbuild.com/Contributing.html#documentation

**Comment by GoaLitiuM** [comment]
Looks like the issue with static linking is more complex than I thought.

It looks like a separate static linker is generally required for building static libraries in meson right now, but D compilers invoke the detected linkers automatically when called with arguments like -lib, which is why the separate static linker is not required. LDC2 doesn't require any external linkers since it comes with built-in LLD linker, but all the compilers prefers other linkers like MSVC linker if they are detected in the system.

What is the proper way to deal with this? Should these linkers need to be implemented in linkers.py with proper detection added into environment.py, or is there a place where DCompilers should insert the required args for static library targets?

File: mesonbuild/environment.py (hunk 6 above)
Comment (GoaLitiuM) [reply mesonbuild/environment.py]: ``detect_cpu_family`` calls the ``detect_windows_arch`` function, but I could call it directly too.

File: mesonbuild/environment.py (hunk 7 above)
Comment (GoaLitiuM) [reply mesonbuild/environment.py]: This line refers to DMD, not LLVM D compiler. All DMD builds (32-bit or 64-bit) on Windows always compiles to 32-bit COFF by default regardless of the environment variables, so we need to know if we are in 64-bit OS, and in MSVC environment for the compiler to target at the right architecture.

As for LLVM D compiler, the compiler does use native MSVC linker if the environment is set up, but it does not use the 32-bit linker when in 32-bit MSVC environment, so this requires adjusting as well.

**Comment by jpakkane** [comment]
The simple solution is to generate a "D linker" class which just does the requisite calls to dmd when invoked.

However.

This breaks down quite badly when mixing languages, for example C and D. I would imagine that in this case you'd want to always use the system static linker.

Could we start by requiring that the "system linker" is always available when compiling D on Windows and just use that? Then in the future we can fix the issue better.

File: mesonbuild/linkers.py [hunk 8]
```diff
@@ -129,3 +129,52 @@ def unix_args_to_native(cls, args):
...
+
+    def get_option_link_args(self, options):
+        return []
+
+    @classmethod
+    def unix_args_to_native(cls, args):
+        return args[:]
+
+    def get_link_debugfile_args(self, targetfile):
+        return []
```
Comment (ignatenkobrain) [inline mesonbuild/linkers.py]: __[Flake8]__

[W292] no newline at end of file

**Comment by GoaLitiuM** [comment]
All the tests should pass now on Linux, but so far only 4/11 passes on Windows. Most of the failures are related to soname linker option getting passed to MSVC linker. I'll take a look at the remaining problems some other day.

File: mesonbuild/compilers/d.py [hunk 9]
```diff
@@ -14,10 +14,13 @@
 
 import os.path, subprocess
 
-from ..mesonlib import EnvironmentException, version_compare
+from ..mesonlib import EnvironmentException, version_compare, is_windows, is_osx
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[F401] '..mesonlib.is_osx' imported but unused

**Comment by GoaLitiuM** [comment]
Unit tests now finish without errors using both LDC2 and DMD as compiler with MSVC linker. 

There seems to be some issues left regarding the dub unit test as the dub package detection method is flawed. The method used for the compiler detection `dub describe <package>` does not detect if the package has been compiled with correct compiler or if compiled at all, but instead the command returns just the build description of the package for the first detected D compiler present in the system, which may be different than the compiler specified by `DC` environment variable. For example, if both DMD and LDC2 are detected by dub, `dub describe <package>` will always return DMD as the detected compiler, even if the package has been previously built with command `dub build <package> --compiler=ldc2`.

Other than that, the PR should be complete for now when it comes to basic D support for Windows. Building pure D projects should work without the need to setup MSVC environment variables, but the linkinking will be faster if the environment is already setup before building the project.

Does this PR require a release notes snippet even though there was no new features added, but only fixes to existing features?

File: mesonbuild/compilers/d.py [hunk 10]
```diff
@@ -215,6 +232,15 @@ def compiles(self, code, env, extra_args=None, dependencies=None, mode='compile'
     def has_multi_arguments(self, args, env):
         return self.compiles('int i;\n', env, extra_args=args)
 
+    def get_target_arch_args(self):
+        # LDC2 on Windows targets to current OS architecture, but
+        # it should follow the target specified by the MSVC toolchain.
+        if is_windows():
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[F821] undefined name 'is_windows'

File: mesonbuild/compilers/d.py [hunk 11]
```diff
@@ -102,6 +109,8 @@ def get_depfile_suffix(self):
         return 'deps'
 
     def get_pic_args(self):
+        if is_windows():
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[F821] undefined name 'is_windows'

File: mesonbuild/compilers/d.py [hunk 12]
```diff
@@ -157,12 +166,20 @@ def get_feature_args(self, kwargs, build_to_src):
...
         return []
 
     def get_std_exe_link_args(self):
         return []
 
+    def gen_import_library_args(self, implibname):
+        return ['-Wl,--out-implib=' + implibname]
+
     def build_rpath_args(self, build_dir, from_dir, rpath_paths, build_rpath, install_rpath):
+        if is_windows():
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[F821] undefined name 'is_windows'

File: mesonbuild/compilers/d.py [hunk 13]
```diff
@@ -409,11 +490,28 @@ def get_linker_search_args(self, dirname):
...
         return d_dmd_buildtype_args[buildtype]
 
     def get_std_shared_lib_link_args(self):
         return ['-shared', '-defaultlib=libphobos2.so']
 
+    def get_target_arch_args(self):
+        # DMD32 and DMD64 on 64-bit Windows defaults to 32-bit (OMF).
+        # Force the target to 64-bit in order to stay consistent
+        # across the different platforms.
+        if is_windows():
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[F821] undefined name 'is_windows'

File: mesonbuild/compilers/d.py [hunk 14]
```diff
@@ -252,22 +282,67 @@ def translate_args_to_nongnu(cls, args):
...
+                continue
             dcargs.append(arg)
 
         return dcargs
 
     def get_debug_args(self, is_debug):
         return clike_debug_args[is_debug]
 
+    def get_crt_args(self, crt_val, buildtype):
+        if not is_windows():
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/d.py]: __[Flake8]__

[F821] undefined name 'is_windows'

**Comment by GoaLitiuM** [comment]
I separated the -soname changes into separate PR #4031 as it also applies to OSX, but otherwise this PR should be good from my part if no other issues raises.

**Comment by nirbheek** [comment]
I noticed that this PR doesn't add CI support for the D language, are you still working on that or is it too difficult to add? Note that in practice, anything that's not being tested on the CI will break very quickly, so it's extremely important to have that.

**Comment by GoaLitiuM** [comment]
I don't know how to configure Appveyor to add this, but DMD and LDC2 for Windows does not require any special installers to work out of box, if the MSVC environment is properly initialized before running the tests.

**Comment by GoaLitiuM** [comment]
I separated some fixes from this PR to #4060 and #4058

File: mesonbuild/backend/ninjabackend.py [hunk 15]
```diff
@@ -2374,7 +2374,7 @@ def generate_link(self, target, outfile, outname, obj_list, linker, extra_args=[
         # options passed on the command-line, in default_options, etc.
         # These have the lowest priority.
         if not isinstance(target, build.StaticLibrary):
-            commands += compilers.get_base_link_args(self.environment.coredata.base_options,
+            commands += compilers.get_base_link_args(self.get_base_options_for_target(target),
```
Comment (GoaLitiuM) [inline mesonbuild/backend/ninjabackend.py]: I'm not sure if this is right, but without this the buildtype option is not accessible in ``get_base_link_args``.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

File: mesonbuild/backend/ninjabackend.py (hunk 15 above)
Comment (nirbheek) [reply mesonbuild/backend/ninjabackend.py]: This is correct, I have a patch sitting locally that fixes this in a bunch of other places too, and I will open a PR for it later. It's blocked on my other PRs.

**Comment by jpakkane** [comment]
Is this ready to merge in its current form?

**Comment by GoaLitiuM** [comment]
It should be ok to merge.

**Comment by jpakkane** [comment]
I need to get it reviewed first ... 😊 

But at least now that 0.47.2 is out and the buildtype branch is in there should be time to get that done.

File: mesonbuild/compilers/compilers.py [hunk 16]
```diff
@@ -459,6 +459,15 @@ def get_base_link_args(options, linker, is_shared_module):
     elif as_needed:
         # -Wl,-dead_strip_dylibs is incompatible with bitcode
         args.append(linker.get_asneeded_args())
+    try:
+        crt_val = options['b_vscrt'].value
+        buildtype = options['buildtype'].value
+        try:
+            args += linker.get_crt_link_args(crt_val, buildtype)
+        except AttributeError:
```
Comment (jpakkane) [inline mesonbuild/compilers/compilers.py]: This should be implemented in the same way as other arg functions. There should be a method of the same name in the base class that returns `[]`. This way there does not need to be any special error handling here.

Comment (GoaLitiuM) [reply mesonbuild/compilers/compilers.py]: I just mirrored how it was done with `get_crt_compile_args`, so surely that should be fixed as well?

**Pull request merged**